feat: json schemas for payloads

heuels opened the pull request:
Hey @gr2m,

In this PR I've added JSON schemas for every payload in `payload-examples`. In order to run a validation, run the following command:

```sh
bin/octokit-payloads.js
```

Closes #136.

dominguezcelada commented on bin/octokit-payloads.js:
```diff
@@ -0,0 +1,21 @@
+  try {
+    const validationResult = validate(event, file);
+    if (!validationResult) {
+      console.error(`X Payload '${filename}' does not match schema`);
```
```suggestion
      console.error(`❌ Payload '${filename}' does not match schema`);
```

dominguezcelada commented on bin/octokit-payloads.js:
```diff
@@ -0,0 +1,21 @@
+    if (!validationResult) {
+      console.error(`X Payload '${filename}' does not match schema`);
+    } else {
+      console.error(`  Payload '${filename}' matches schema`);
```
```suggestion
      console.log(`✅ Payload '${filename}' matches schema`);
```

dominguezcelada commented on bin/octokit-payloads.js:
```diff
@@ -0,0 +1,21 @@
+const fs = require('fs');
+const validate = require('../payload-schemas');
+
+fs.readdirSync('./payload-examples/api.github.com').forEach(filename => {
+  const [event, ...unused] = filename.split('.');
+  const fileBuffer = fs.readFileSync(`./payload-examples/api.github.com/${filename}`);
```
[💡 suggestion]: Not sure if there would be performance improvements on this but, do you think would be possible to use asynchronous `readdir` (https://nodejs.org/api/fs.html#fs_fs_readdir_path_options_callback) or `fs.promises.readdir` (https://nodejs.org/api/fs.html#fs_fspromises_readdir_path_options)?

heuels replied:
Thanks for the suggestion!

I've seen only synchronous methods throughout the project and just went with it. Personally, I don't think that there would be any measurable difference between synchronous and asynchronous methods here, but I'd be OK with replacing them with asynchronous ones should you ask me to :)

dominguezcelada replied:
Sure! If you think it would not be an impact and the repo is using sync everywhere... to me it's fine. It was just a suggestion 😄

gr2m replied:
for build scripts like this I prefer synchronous methods, they are easier to reason about. If this could would run inside e.g. a server request handler, then we wouldn't want to block the event loop, and the async methods should be used. But that doesn't apply here

dominguezcelada commented on payload-schemas/index.js:
```diff
@@ -0,0 +1,69 @@
+const Ajv = require('ajv');
```
Could it be possible you forgot to commit this devDependency in `package.json` + `package-lock.json`?

heuels replied:
Oops, definitely so. Thanks!

dominguezcelada replied:
Oh! It was already in the project! Didn't notice that. Sorry 🙏🏽

dominguezcelada commented:
Amazing the work you did with the huge amount of schemas to migrate @heuels ! I left some few suggestions but the code looks good to me. Some questions I have are:

- Do you think it would be interesting to test somehow what you did here? Not sure if you think it could be interesting or valuable.
- Should this new code validation have to be integrated somehow to run it under `npm run build` script in package.json and the Github Action which auto-maintains it? Or maybe this would be in a different PR? https://github.com/octokit/webhooks/blob/6ea05de37e22647729d79ee9e862364e8cd371c7/package.json#L21 
https://github.com/octokit/webhooks/blob/6ea05de37e22647729d79ee9e862364e8cd371c7/.github/workflows/update.yml#L18
- Which tool did you use to generate the schemas? https://jsonschema.net/ ? Manual? If it was automated, maybe we should try to run it to maintain schemas maybe?

heuels commented on payload-schemas/index.js:
```diff
@@ -0,0 +1,69 @@
+ajv.addSchema(require('./schemas/team_add.schema.json'));
+ajv.addSchema(require('./schemas/watch.schema.json'));
+ajv.addSchema(require('./schemas/workflow_dispatch.schema.json'));
+ajv.addSchema(require('./schemas/workflow_run.schema.json'));
```
Oh yes, that's a fantastic idea, I will do it now.

It's just that I've been adding schemas to this long list one by one so it didn't seem like a monstrosity to me at the time 😀

dominguezcelada replied:
haha, I imagined it was that, you were incrementally working on them. That's why review's are cool, someone external can see out of the box sometimes :)

dominguezcelada commented on package.json:
```diff
@@ -8,7 +8,7 @@
   "devDependencies": {
     "@octokit/request": "^5.0.0",
-    "ajv": "^6.12.2",
+    "ajv": "^6.12.5",
```
Oh! It was already in the project! Didn't notice that. Sorry 🙏🏽

heuels commented:
Hey @dominguezcelada, thank you so much for the review comments!

> Which tool did you use to generate the schemas? jsonschema.net ? Manual? If it was automated, maybe we should try to run it to maintain schemas maybe?

I'm sorry to say it was mostly manual. I copied payload examples from this repository, pasted them to jsonschema.net, enriched the result with enum values from documentation and finally manually tweaked them until schema matched all of the example payloads.

> Should this new code validation have to be integrated somehow to run it under `npm run build` script in package.json and the Github Action which auto-maintains it? Or maybe this would be in a different PR?

Hmm, I've looked through the code once again and I think that I'll move the code from `bin/octokit-payloads.js` to `test.js` so that validation runs every time `npm test` is run. Auto-maintaining the schemas is a whole other challenge 🙂 

> Do you think it would be interesting to test somehow what you did here? Not sure if you think it could be interesting or valuable.

I've mostly added schemas, they're not really something that's worth testing. But I'm open to suggestions! 🙂

heuels commented:
Oh, it seems that I haven't merged main branch into mine for quite a while – there are new payloads which don't match the schema. I will look into it in the evening.

dominguezcelada commented:
> Oh, it seems that I haven't merged main branch into mine for quite a while – there are new payloads which don't match the schema. I will look into it in the evening.

No problem! When you have time! And if you need help let us know :)

dominguezcelada commented:
> Hey @dominguezcelada, thank you so much for the review comments!
> 
> > Which tool did you use to generate the schemas? jsonschema.net ? Manual? If it was automated, maybe we should try to run it to maintain schemas maybe?
> 
> I'm sorry to say it was mostly manual. I copied payload examples from this repository, pasted them to jsonschema.net, enriched the result with enum values from documentation and finally manually tweaked them until schema matched all of the example payloads.

If it was generated mostly manual it has no sense to automate it at this stage. Let's see what @gr2m thinks about this. 

> > Should this new code validation have to be integrated somehow to run it under `npm run build` script in package.json and the Github Action which auto-maintains it? Or maybe this would be in a different PR?
> 
> Hmm, I've looked through the code once again and I think that I'll add `bin/octokit-payloads.js` to `test` script in `package.json` so that validation runs every time `npm test` is run. Auto-maintaining the schemas is a whole other challenge 🙂

Sounds good to me. Checking Github Action, I think we don't need to touch anything else there. Agree on that @gr2m ?

> > Do you think it would be interesting to test somehow what you did here? Not sure if you think it could be interesting or valuable.
> 
> I've mostly added schemas, they're not really something that's worth testing. But I'm open to suggestions! 🙂

I was thinking about making sure it exits(1) when is expected to when running `bin/octokit-payloads.js`. To not block this maybe I can draft a PR with tests after this is merged. As you mention, there is not that much logic to test.

Let's see what @gr2m thinks.

gr2m commented:
I can do a more thorough review next week

> > I'm sorry to say it was mostly manual. I copied payload examples from this repository, pasted them to jsonschema.net, enriched the result with enum values from documentation and finally manually tweaked them until schema matched all of the example payloads.
> 
> If it was generated mostly manual it has no sense to automate it at this stage. Let's see what @gr2m thinks about this.

It might make sense to generate the schema initially based on the examples we have, but moving forward we will have to maintain it manually. We should however test the schema against the example payloads we have, to make sure it's valid. Then whenever the schema needs to be changed / extended, we can add a new payload as a means of testing the schema and to prevent regressions

wolfy1339 commented on payload-schemas/schemas/common/installation.schema.json:
```diff
@@ -0,0 +1,14 @@
+    }
+  },
+  "additionalProperties": false
+}
```
Make sure to add the schema files to the files passed by the linter.
There needs to be a new-line here

wolfy1339 commented on payload-schemas/schemas/common/organization.schema.json:
```diff
@@ -0,0 +1,44 @@
+{
```
This seems to not contain all the fields.

You can cross-reference https://api.github.com/orgs/Octocoders and make sure the `company`, `blog`, `email`, and `twitter_username` are `null | string`

wolfy1339 commented on payload-schemas/schemas/common/repository.schema.json:
```diff
@@ -0,0 +1,406 @@
+{
```
There's an `organization` property here, with a type of `string`
Should be optional, just like the organization object itself.

wolfy1339 commented on payload-schemas/schemas/common/organization.schema.json:
```diff
@@ -0,0 +1,44 @@
+      "type": "string"
+    },
+    "description": {
+      "type": "string"
```
This can also be `null`

wolfy1339 reviewed (changes requested):
Make sure to pass the files through the linter and just address my comments on the common schemas.

Otherwise, it looks good. Thanks for the work

wolfy1339 commented on package.json:
```diff
@@ -20,9 +20,10 @@
     "pretest": "npm run -s lint",
-    "test": "npm run build && node test.js",
+    "test": "npm run build && npm run validate && node test.js",
     "lint": "prettier --check '{bin,lib}/*.js' '*.{js,md}' 'payload-examples/**/*.json' *.md package.json index.json",
```
```suggestion
    "lint": "prettier --check '{bin,lib}/*.js' '*.{js,md}' 'payload-examples/**/*.json' 'payload-schemas/*.js' 'payload-schemas/schemas/**/*.json' *.md package.json index.json",
```

wolfy1339 commented on package.json:
```diff
@@ -20,9 +20,10 @@
+    "test": "npm run build && npm run validate && node test.js",
     "lint": "prettier --check '{bin,lib}/*.js' '*.{js,md}' 'payload-examples/**/*.json' *.md package.json index.json",
-    "lint:fix": "prettier --write '{bin,lib}/*.js' '*.{js,md}' 'payload-examples/**/*.json' *.md package.json index.json"
+    "lint:fix": "prettier --write '{bin,lib}/*.js' '*.{js,md}' 'payload-examples/**/*.json' *.md package.json index.json",
```
```suggestion
    "lint:fix": "prettier --write '{bin,lib}/*.js' '*.{js,md}' 'payload-examples/**/*.json' 'payload-schemas/*.js' 'payload-schemas/schemas/**/*.json' *.md package.json index.json",
```

dominguezcelada commented:
Quoting @gr2m suggestion from another PR:

> Once we have the schema, we can use it as the source of truth. We can continue to use the example payloads as a means to test the schema, and each time we find a problem with the schema, we should add an example payload to avoid future regressions

https://github.com/octokit/webhooks.js/issues/266#issuecomment-707252262

gr2m commented:
@heuels do you still plan on finishing up the PR?

heuels commented:
@gr2m, sorry, this fell off my radar lately. I will try to finish it up this weekend.

G-Rath commented:
Is there a particular reason that `additionalProperties` is set to `true` on what looks to be the majority of the schemas?

Ideally it should be `false` everywhere unless there really is an object whose keys are defined dynamically based on external sources - while the `required` field will ensure you're not missing properties, it'll result in weaker types and checks (i.e typescript generated from the schemas will have to stick `[k: string]: unknown;` on the type meaning TS won't error if you reference properties that don't exist).

wolfy1339 commented:
@heuels Are you still interested in finishing this? There's another user that could contribute to finish this if you don't plan on finishing this

heuels commented:
@wolfy1339, once again, sorry for the delay. I've been on and off on this task for a couple of weeks, just did not push the changes here. But from what I can see all the schemas are now in place and ready for another review.

@G-Rath, thanks, that's valid point and I fixed that. The only place left with `additionalProperties: true` is the `inputs` object in `workflow_dispatch` payload.

gr2m commented on payload-schemas/schemas/repository_dispatch.schema.json:
```diff
@@ -0,0 +1,49 @@
+    },
+    "client_payload": {
+      "type": "object",
+      "required": ["unit", "integration"],
+      "properties": {
+        "unit": {
+          "type": "boolean"
+        },
+        "integration": {
+          "type": "boolean"
+        }
+      },
+      "additionalProperties": false
```
`client_payload` is always different. The `unit` and `integration` properties are probably just coming from an example, we can remove them. And `"additionalProperties"` should be true
```suggestion
      "properties": {},
      "additionalProperties": true
```

heuels replied:
Thanks, I missed that somehow. Fixed now.

gr2m commented on bin/octokit-payloads.js:
```diff
@@ -0,0 +1,32 @@
+
+if (hasErrors) {
+  process.exit(1);
+}
```
should we maybe rename the file to `bin/validate-payload-examples.js`

heuels replied:
Done. And I added another script `bin/octokit-schema.js` as well that generates a dereferenced JSON schema.

gr2m commented on bin/octokit-payloads.js:
```diff
@@ -0,0 +1,32 @@
+  .filter((filename) => filename.endsWith(".json"))
+  .forEach((filename) => {
+    const [event, ...unused] = filename.split(".");
+    const fileBuffer = fs.readFileSync(`${payloads}/${filename}`);
+    const file = JSON.parse(fileBuffer.toString());
```
you can just use `require` for JSON files

```suggestion
    const file = require(`../payload-examples/api.github.com/${filename}`);
```

gr2m commented on bin/octokit-payloads.js:
```diff
@@ -0,0 +1,32 @@
+fs.readdirSync(payloads)
+  .filter((filename) => filename.endsWith(".json"))
+  .forEach((filename) => {
+    const [event, ...unused] = filename.split(".");
```
No need for the `unused` part

```suggestion
    const [event] = filename.split(".");
```

gr2m commented on package.json:
```diff
@@ -20,9 +20,10 @@
-    "lint": "prettier --check '{bin,lib}/*.js' '*.{js,md}' 'payload-examples/**/*.json' *.md package.json index.json",
-    "lint:fix": "prettier --write '{bin,lib}/*.js' '*.{js,md}' 'payload-examples/**/*.json' *.md package.json index.json"
+    "test": "npm run build && npm run validate && node test.js",
+    "lint": "prettier --check '{bin,lib}/*.js' '*.{js,md}' 'payload-examples/**/*.json' 'payload-schemas/schemas/*.json' 'payload-schemas/schemas/common/*.json' 'payload-schemas/*.js' *.md package.json index.json",
+    "lint:fix": "prettier --write '{bin,lib}/*.js' '*.{js,md}' 'payload-examples/**/*.json' 'payload-schemas/schemas/*.json' 'payload-schemas/schemas/common/*.json' 'payload-schemas/*.js' *.md package.json index.json",
```
`'payload-examples/**/*.json'` should find all JSON files in all folders within `payload-examples`

```suggestion
    "lint": "prettier --check '{bin,lib}/*.js' '*.{js,md}' 'payload-examples/**/*.json' 'payload-schemas/*.js' *.md package.json index.json",
    "lint:fix": "prettier --write '{bin,lib}/*.js' '*.{js,md}' 'payload-examples/**/*.json' 'payload-schemas/*.js' *.md package.json index.json",
```

gr2m reviewed (changes requested):
This is really great, thanks so much @heuels! I found one schema that needs to be updated and some change suggestions.

We also need to update the `"files"` key in the package.json. I also wonder if we should generate a de-referenced schema file, which we would only build before publish, not commit to the repository. That would make it easier for consumers, as they wouldn't need to stitch it together themselves? 

Could you also update the README.md file with instructions how to use the schemas?

I also think that one schema per event probably won't suffice, I think there are several cases where certain keys are present for some actions but not others. But I'd say that's something we can figure out once we land this pull request

heuels commented:
@gr2m, hi!

> I also wonder if we should generate a de-referenced schema file, which we would only build before publish, not commit to the repository.

I've added a dereferenced schema with a build script, added it to `"files"` property in `package.json` and `.gitignore`.

> I also think that one schema per event probably won't suffice, I think there are several cases where certain keys are present for some actions but not others. But I'd say that's something we can figure out once we land this pull request

Yes, I've come to the same conclusion. We'll need separate schemas for some of the action types.

gr2m commented on bin/validate-payload-examples.js:
```diff
@@ -0,0 +1,32 @@
+
+if (hasErrors) {
+  process.exit(1);
+}
```
What do you think about combining `bin/validate-payload-examples.js` and `test.js`? I think they are pretty smilar in scope now? We can do that in a follow up PR, I'll try to get this merged as is

heuels replied:
Sure, they really are pretty similar. I'll open a PR with this change once this PR is merged.

gr2m commented on payload-schemas/schemas/repository.schema.json:
```diff
@@ -0,0 +1,41 @@
+        "privatized"
+      ]
+    },
+    "changes": {
+      "type": "object",
+      "properties": {},
+      "additionalProperties": true
+    },
```
I'd say let's be more strict/explicit. We can expand the schema as we add more payload examples, see the discussion at https://github.com/octokit/webhooks/pull/235

```suggestion
    "changes": {
      "type": "object", 
      "properties": {
        "description": {
          "type": "object", 
          "properties": {
            "from": {
              "type": "string"
            }
          },
          "additionalProperties": false
        }
      },
      "additionalProperties": false
    }
```

heuels replied:
The discussion moved forward and I happened to overlook this, sorry. Anyways, it looks like that with the effort of self-updating schemas underway it'd be OK to leave it as-is. Please let me know what you think.

gr2m commented:
Other than my last comment https://github.com/octokit/webhooks/pull/186#discussion_r546283102, this is good to go. Let me know what you think about my change suggestion

G-Rath commented on payload-schemas/schemas/common/installation.schema.json:
```diff
@@ -0,0 +1,14 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema",
+  "$id": "common/installation.schema.json",
+  "type": ["object", "null"],
```
I just noticed these - I think it would be better to use `oneOf` at the reference site to mark possible nulls rather than to have this `null` here.

My reasoning is that right now the type is this:

```

type GithubInstallation= { ... } | null;

interface MyPayload {
  installation: GithubInstallation;
}
```

That means theres no way to reference the object type for `Installation` without always doing some extra work:

```
const describeInstallation = (installation: Exclude<GithubInstallation, null>) => {
  // 
};
```

Even if the property could be `null` everywhere the type is used, it's more to indicate that at the property-level rather than the type-level.

gr2m replied:
makes sense, can you make the change?

G-Rath replied:
Yup, will do tomorrow.

G-Rath commented:
I've created an Azure Function that uses these schemas, comparing incoming webhooks and reports if they're valid or not.

I'll push the source code up tomorrow, but you can use it by setting up a webhook pointing [here](sls-ause-dev-gh-webhooks-schema-checker.azurewebsites.net/api/github) - the function returns json that includes the errors if it's not valid, so you can inspect the Recent Events in the Webhook Settings page to see what's happening

![image](https://user-images.githubusercontent.com/3151613/102853144-9fa9a880-4484-11eb-975c-0db22a5e168b.png)

Including this in are scripts for generating what I think should be the new layout for schemas + the types for the schemas.

<details>
<summary>this the folder structure of the new schema layout</summary>

```
src/schemas/
├── check_run
│   ├── completed.d.ts
│   ├── completed.schema.json
│   ├── created.d.ts
│   ├── created.schema.json
│   ├── requested_action.d.ts
│   ├── requested_action.schema.json
│   ├── rerequested.d.ts
│   └── rerequested.schema.json
├── check_suite
│   ├── completed.d.ts
│   ├── completed.schema.json
│   ├── requested.d.ts
│   ├── requested.schema.json
│   ├── rerequested.d.ts
│   └── rerequested.schema.json
├── code_scanning_alert
│   ├── appeared_in_branch.d.ts
│   ├── appeared_in_branch.schema.json
│   ├── closed_by_user.d.ts
│   ├── closed_by_user.schema.json
│   ├── created.d.ts
│   ├── created.schema.json
│   ├── fixed.d.ts
│   ├── fixed.schema.json
│   ├── reopened.d.ts
│   ├── reopened.schema.json
│   ├── reopened_by_user.d.ts
│   └── reopened_by_user.schema.json
├── commit_comment
│   ├── created.d.ts
│   └── created.schema.json
├── common
│   ├── installation.d.ts
│   ├── installation.schema.json
│   ├── organization.d.ts
│   ├── organization.schema.json
│   ├── repository.d.ts
│   ├── repository.schema.json
│   ├── user.d.ts
│   └── user.schema.json
├── content_reference
│   ├── created.d.ts
│   └── created.schema.json
├── create
│   ├── event.d.ts
│   └── event.schema.json
├── delete
│   ├── event.d.ts
│   └── event.schema.json
├── deploy_key
│   ├── created.d.ts
│   ├── created.schema.json
│   ├── deleted.d.ts
│   └── deleted.schema.json
├── deployment
│   ├── created.d.ts
│   └── created.schema.json
├── deployment_status
│   ├── created.d.ts
│   └── created.schema.json
├── fork
│   ├── event.d.ts
│   └── event.schema.json
├── github_app_authorization
│   ├── revoked.d.ts
│   └── revoked.schema.json
├── gollum
│   ├── event.d.ts
│   └── event.schema.json
├── installation
│   ├── created.d.ts
│   ├── created.schema.json
│   ├── deleted.d.ts
│   └── deleted.schema.json
├── installation_repositories
│   ├── added.d.ts
│   ├── added.schema.json
│   ├── removed.d.ts
│   └── removed.schema.json
├── integration_installation
│   ├── created.d.ts
│   └── created.schema.json
├── integration_installation_repositories
│   ├── added.d.ts
│   └── added.schema.json
├── issue_comment
│   ├── created.d.ts
│   ├── created.schema.json
│   ├── deleted.d.ts
│   ├── deleted.schema.json
│   ├── edited.d.ts
│   └── edited.schema.json
├── issues
│   ├── assigned.d.ts
│   ├── assigned.schema.json
│   ├── closed.d.ts
│   ├── closed.schema.json
│   ├── deleted.d.ts
│   ├── deleted.schema.json
│   ├── demilestoned.d.ts
│   ├── demilestoned.schema.json
│   ├── edited.d.ts
│   ├── edited.schema.json
│   ├── labeled.d.ts
│   ├── labeled.schema.json
│   ├── locked.d.ts
│   ├── locked.schema.json
│   ├── milestoned.d.ts
│   ├── milestoned.schema.json
│   ├── opened.d.ts
│   ├── opened.schema.json
│   ├── pinned.d.ts
│   ├── pinned.schema.json
│   ├── reopened.d.ts
│   ├── reopened.schema.json
│   ├── transferred.d.ts
│   ├── transferred.schema.json
│   ├── unassigned.d.ts
│   ├── unassigned.schema.json
│   ├── unlabeled.d.ts
│   ├── unlabeled.schema.json
│   ├── unlocked.d.ts
│   ├── unlocked.schema.json
│   ├── unpinned.d.ts
│   └── unpinned.schema.json
├── label
│   ├── created.d.ts
│   ├── created.schema.json
│   ├── deleted.d.ts
│   ├── deleted.schema.json
│   ├── edited.d.ts
│   └── edited.schema.json
├── marketplace_purchase
│   ├── cancelled.d.ts
│   ├── cancelled.schema.json
│   ├── changed.d.ts
│   ├── changed.schema.json
│   ├── purchased.d.ts
│   └── purchased.schema.json
├── member
│   ├── added.d.ts
│   ├── added.schema.json
│   ├── edited.d.ts
│   ├── edited.schema.json
│   ├── removed.d.ts
│   └── removed.schema.json
├── membership
│   ├── added.d.ts
│   ├── added.schema.json
│   ├── removed.d.ts
│   └── removed.schema.json
├── meta
│   ├── deleted.d.ts
│   └── deleted.schema.json
├── milestone
│   ├── closed.d.ts
│   ├── closed.schema.json
│   ├── created.d.ts
│   ├── created.schema.json
│   ├── deleted.d.ts
│   ├── deleted.schema.json
│   ├── edited.d.ts
│   ├── edited.schema.json
│   ├── opened.d.ts
│   └── opened.schema.json
├── org_block
│   ├── blocked.d.ts
│   ├── blocked.schema.json
│   ├── unblocked.d.ts
│   └── unblocked.schema.json
├── organization
│   ├── deleted.d.ts
│   ├── deleted.schema.json
│   ├── member_added.d.ts
│   ├── member_added.schema.json
│   ├── member_invited.d.ts
│   ├── member_invited.schema.json
│   ├── member_removed.d.ts
│   ├── member_removed.schema.json
│   ├── renamed.d.ts
│   └── renamed.schema.json
├── package
│   ├── published.d.ts
│   ├── published.schema.json
│   ├── updated.d.ts
│   └── updated.schema.json
├── page_build
│   ├── event.d.ts
│   └── event.schema.json
├── ping
│   ├── event.d.ts
│   └── event.schema.json
├── project
│   ├── closed.d.ts
│   ├── closed.schema.json
│   ├── created.d.ts
│   ├── created.schema.json
│   ├── deleted.d.ts
│   ├── deleted.schema.json
│   ├── edited.d.ts
│   ├── edited.schema.json
│   ├── reopened.d.ts
│   └── reopened.schema.json
├── project_card
│   ├── converted.d.ts
│   ├── converted.schema.json
│   ├── created.d.ts
│   ├── created.schema.json
│   ├── deleted.d.ts
│   ├── deleted.schema.json
│   ├── edited.d.ts
│   ├── edited.schema.json
│   ├── moved.d.ts
│   └── moved.schema.json
├── project_column
│   ├── created.d.ts
│   ├── created.schema.json
│   ├── deleted.d.ts
│   ├── deleted.schema.json
│   ├── edited.d.ts
│   ├── edited.schema.json
│   ├── moved.d.ts
│   └── moved.schema.json
├── public
│   ├── event.d.ts
│   └── event.schema.json
├── pull_request
│   ├── assigned.d.ts
│   ├── assigned.schema.json
│   ├── closed.d.ts
│   ├── closed.schema.json
│   ├── edited.d.ts
│   ├── edited.schema.json
│   ├── labeled.d.ts
│   ├── labeled.schema.json
│   ├── locked.d.ts
│   ├── locked.schema.json
│   ├── opened.d.ts
│   ├── opened.schema.json
│   ├── ready_for_review.d.ts
│   ├── ready_for_review.schema.json
│   ├── reopened.d.ts
│   ├── reopened.schema.json
│   ├── review_request_removed.d.ts
│   ├── review_request_removed.schema.json
│   ├── review_requested.d.ts
│   ├── review_requested.schema.json
│   ├── synchronize.d.ts
│   ├── synchronize.schema.json
│   ├── unassigned.d.ts
│   ├── unassigned.schema.json
│   ├── unlabeled.d.ts
│   ├── unlabeled.schema.json
│   ├── unlocked.d.ts
│   └── unlocked.schema.json
├── pull_request_review
│   ├── submitted.d.ts
│   └── submitted.schema.json
├── pull_request_review_comment
│   ├── created.d.ts
│   └── created.schema.json
├── push
│   ├── event.d.ts
│   └── event.schema.json
├── release
│   ├── created.d.ts
│   ├── created.schema.json
│   ├── deleted.d.ts
│   ├── deleted.schema.json
│   ├── edited.d.ts
│   ├── edited.schema.json
│   ├── prereleased.d.ts
│   ├── prereleased.schema.json
│   ├── published.d.ts
│   ├── published.schema.json
│   ├── released.d.ts
│   ├── released.schema.json
│   ├── unpublished.d.ts
│   └── unpublished.schema.json
├── repository
│   ├── archived.d.ts
│   ├── archived.schema.json
│   ├── created.d.ts
│   ├── created.schema.json
│   ├── deleted.d.ts
│   ├── deleted.schema.json
│   ├── edited.d.ts
│   ├── edited.schema.json
│   ├── privatized.d.ts
│   ├── privatized.schema.json
│   ├── publicized.d.ts
│   ├── publicized.schema.json
│   ├── renamed.d.ts
│   ├── renamed.schema.json
│   ├── transferred.d.ts
│   ├── transferred.schema.json
│   ├── unarchived.d.ts
│   └── unarchived.schema.json
├── repository_dispatch
│   ├── event.d.ts
│   └── event.schema.json
├── repository_import
│   ├── event.d.ts
│   └── event.schema.json
├── repository_vulnerability_alert
│   ├── create.d.ts
│   ├── create.schema.json
│   ├── dismiss.d.ts
│   ├── dismiss.schema.json
│   ├── resolve.d.ts
│   └── resolve.schema.json
├── security_advisory
│   ├── performed.d.ts
│   ├── performed.schema.json
│   ├── published.d.ts
│   ├── published.schema.json
│   ├── updated.d.ts
│   └── updated.schema.json
├── sponsorship
│   ├── cancelled.d.ts
│   ├── cancelled.schema.json
│   ├── created.d.ts
│   ├── created.schema.json
│   ├── edited.d.ts
│   ├── edited.schema.json
│   ├── pending_cancellation.d.ts
│   ├── pending_cancellation.schema.json
│   ├── pending_tier_change.d.ts
│   ├── pending_tier_change.schema.json
│   ├── tier_changed.d.ts
│   └── tier_changed.schema.json
├── star
│   ├── created.d.ts
│   ├── created.schema.json
│   ├── deleted.d.ts
│   └── deleted.schema.json
├── status
│   ├── event.d.ts
│   └── event.schema.json
├── team
│   ├── added_to_repository.d.ts
│   ├── added_to_repository.schema.json
│   ├── created.d.ts
│   ├── created.schema.json
│   ├── deleted.d.ts
│   ├── deleted.schema.json
│   ├── edited.d.ts
│   ├── edited.schema.json
│   ├── removed_from_repository.d.ts
│   └── removed_from_repository.schema.json
├── team_add
│   ├── event.d.ts
│   └── event.schema.json
├── watch
│   ├── started.d.ts
│   └── started.schema.json
├── workflow_dispatch
│   ├── event.d.ts
│   └── event.schema.json
└── workflow_run
    ├── event.d.ts
    └── event.schema.json

53 directories, 310 files
```
</details>

Right now I'm creating a `oneOf` schema inline at validation time:

```
this._validator = this._ajv.compile({
  $schema: 'http://json-schema.org/draft-07/schema',
  $id: eventName,
  oneOf: schemaIds.map(id => ({ $ref: id }))
});
```

but it'd be easy enough to extend the copy script to also create that and write it to disk.

(this was born out of the conversation me & @gr2m has [here](https://github.com/octokit/webhooks/pull/235#issuecomment-748519592) - the actual function didn't take all that long, but I decided it was a good excuse to try out Azure Functions, which lead me on a rabbit hole and learning that the serverless ecosystem is very aws-first so I had to step on a few rakes before I got it all working 😕)

wolfy1339 commented:
Hey @G-Rath it seems your application is returning `500`

G-Rath commented:
@wolfy1339 🤦 it'll be because of the secret - it should be "thisismysecret".

wolfy1339 commented:
@G-Rath It seems like i've already found a discrepancy between the schema and the webhooks, or it's just errors on your part.
```json
{
  "summary": "oh dear, there were errors :/",
  "errors": [
    {
      "keyword": "type",
      "dataPath": "/description",
      "schemaPath": "#/properties/description/type",
      "params": {
        "type": "null"
      },
      "message": "should be null"
    },
    {
      "keyword": "oneOf",
      "dataPath": "",
      "schemaPath": "#/oneOf",
      "params": {
        "passingSchemas": null
      },
      "message": "should match exactly one schema in oneOf"
    }
  ]
}
```

G-Rath commented:
@wolfy1339 that's a discrepancy: 

https://github.com/octokit/webhooks/blob/80db882a7cfdd32d4a549f677ba32defea9d34d4/payload-schemas/schemas/create.schema.json#L25-L27

(once you've got to the schema validating code, it should all be discrepancies rather than errors on my end as the code is mainly just about loading the schemas and passing the data to `ajv` 🤞)

wolfy1339 commented:
That's great news! I guess the example payloads need to be updated

gr2m commented:
is the source code for your Azure function on GitHub, so we can discuss? If not we can create a separate issue here.

One idea I have is that your app could directly create pull requests submitting a payload that failed to validate. In order to not create endless pull requests, we could create a hash from event name + action + validation error, and use that as the branch name for the pull request. If it already exists, no need to create it again.

Once we have that setup I could write a script which would add setup a webhook manually across tons of repositories. I'd start with the @octokit repos and some personal ones

G-Rath commented:
Yup, it's [here](https://github.com/G-Rath/compare-gh-webhook-to-schema-function) - it's even got its first [issue](https://github.com/G-Rath/compare-gh-webhook-to-schema-function/issues/1) thanks to @wolfy1339 😄 (which I plan to action shortly - I've been deep diving into type generation for the last hour or so 😅)

> One idea I have is that your app could directly create pull requests submitting a payload that failed to validate.

For sure - I held off doing that for now just because of the number of schemas we've got (or lack thereof) and not wanting to spam a repo out of the blue, but that sort of automation is definitely one of the medium-long term feature goals.

wolfy1339 commented:
This requires #236, since I added a new property to the schema which is found in the updated payload examples.
@heuels Can you update with the latest changes from master and update the schemas accordingly. Also, please address the review comments.
Thanks for the work you put in to this

G-Rath commented:
btw also on my todo-list is to have the repo self-update & redeploy when the schemas here are changed.

Also @gr2m @heuels @wolfy1339 [this](https://github.com/G-Rath/compare-gh-webhook-to-schema-function/tree/main/src/schemas) is the new structure I'm proposing for the schema files - would appreciate your thoughts :)

It's generated via a [script](https://github.com/G-Rath/compare-gh-webhook-to-schema-function/blob/main/scripts/copy-schemas.ts) so it shouldn't be a problem to work on this PR as is and then when ready just update it with the new structure if we're happy with that.

Right now I'm working on generating the types, which live next to the schema files. It works pretty well, just figuring out how to handle de-referencing.

gr2m commented:
> btw also on my todo-list is to have the repo self-update & redeploy when the schemas here are changed.

you can install this app in order to receive repository dispatch events for each new release in this repository:
https://github.com/apps/octokit-release-notifier/

gr2m commented:
> Right now I'm working on generating the types, which live next to the schema files. It works pretty well, just figuring out how to handle de-referencing.

I would you need de-referencing? Could you use the de-referenced schema as input? That's what we do with the OpenAPI spec -> TypeScript in https://github.com/octokit/openapi-types.ts

G-Rath commented:
I've finished writing the type generation script 🎉 

You can find it [here](https://github.com/G-Rath/compare-gh-webhook-to-schema-function/pull/4).

Now that I've got that working nicely, I'm keen to pick up working on these schemas - what'd be the best/prefered way to do that since I'm not a maintainer (and so can't push to this PR directly)? I could branch off this PR and make a new one, or otherwise... do something else 😅

wolfy1339 commented:
I'd branch off of this PR, or we could land this PR and then we can do all the optimisations afterwards

G-Rath commented:
tbh the thing with landing is that it'd trigger a release (and that there are PR bots watching for those releases on other repos) - maybe we should have a new `next`-style branch on this repo to PR into instead? i.e `add-schemas`.

A new release isn't the end of the world, but just aware that there will likely be a big flurry of optimizations (including a whole directory structure shuffle) that'll create a lot of noise.

heuels commented:
@wolfy1339, @G-Rath, I'll implement support for recent changes in payloads tomorrow and it looks like this PR will be ready for merge then.

I don't think that a new release that will follow the merging of this PR will really matter, because there'll be no significant changes apart from the added schema file.

wolfy1339 commented:
Usually what would happen is we would create a new branch off of master that would remain the stable current version, and the master branch would be the pre-release version.

That's what happens with major version bumps anyways.

G-Rath commented:
@wolfy1339 that'd work fine too. I'm just keen to get going on some of the improvements on my list, and for making more schemas from things :D

Ultimately these schemas are not being used by anything yet, and they're primarily being done for type generation anyway - I've got no problem with how you want to approach this so long as I can make some PRs to start landing some things here.

gr2m commented:
@G-Rath I've invited you to the repo: https://github.com/octokit/webhooks/invitations. That'll make it easier to collaborate on the schema moving forward :)

G-Rath commented:
@gr2m oh awesome thanks, that'll make things easier!

@heuels if it's alright with you, I'll push some of my minor changes to your branch i.e [adjusting the `[object, null]` type on the common schemas](https://github.com/octokit/webhooks/pull/186#discussion_r546317207), adding `title` properties in the right places for the type generation, etc. 

For now I'll leave the work of implementing support for the recent changes to payloads for you since you're planning to do that tomorrow and might have already started, but let me know if you'd like a hand!

heuels commented:
@G-Rath, I've included changes from the `master` branch and adapted the schemas. And it's totally alright to push to my branch 🙂 

@wolfy1339, it looks like some of the payloads that you merged recently had a different structure from the ones that were in the repo. I've taken the liberty of modifying them to match the rest, but please let me know if it was not an error. Example here: https://github.com/octokit/webhooks/pull/186/commits/880d7dfadf7e8f04f179dcb463a33ee8cdbaa778#diff-b2c99683454a9b3f213462254777b05bbbcf6d594212325105ccdc139b792ab1

wolfy1339 commented:
By structure, what do you mean?

If it's just a different ordering, then it shouldn't matter, it's fine to leave them as-is. If it's new properties, then the schemas need updating, I took the payloads from actual webhooks

heuels commented:
@wolfy1339, yes, by structure. The objects had two top-level keys, `event` with the webhook name and `payload`, which contained actual webhook payload – so I extracted it to the root.

wolfy1339 commented:
I see now. 
Can you extract those changes into another commit? It will make it more obvious why the change is there

heuels commented:
Sure thing. I could also open a PR to fix these in `master`, if it's OK with you.

wolfy1339 commented:
Of course! Go ahead and submit a PR fixing them

wolfy1339 commented on package.json:
```diff
@@ -7,32 +7,36 @@
-    "build": "node bin/octokit-webhooks.js check --cached",
+    "build": "npm run build:schema && npm run build:webhooks",
+    "build:schema": "node bin/octokit-schema.js",
+    "build:webhooks": "node bin/octokit-webhooks.js check --cached",
```
To update the `index.json` file, the command is actually `node bin/octokit-webhooks.js update`.

The current `build` command, only updates the cache of webhooks docs and is used by this GitHub workflow: https://github.com/octokit/webhooks/blob/9b8cac417e14b5ff5ba7c456b44d431db305fe1c/.github/workflows/update.yml

wolfy1339 replied:
```suggestion
    "build:webhooks": "node bin/octokit-webhooks.js update",
```

wolfy1339 commented on README.md:
```diff
@@ -225,20 +225,38 @@ Example webhook definition
-[`index.json`](index.json) file is
-generated by [`bin/octokit-webhooks.js`](bin/octokit-webhooks.js). Run
-`node bin/octokit-webhooks --usage` for instructions.
+This package updates itself using a daily cronjob running on GitHub Actions. The [`index.json`](index.json) file is generated by [`bin/octokit-webhooks.js`](bin/octokit-webhooks.js). Run `node bin/octokit-webhooks --usage` for instructions. After the update is complete, run `npm run build` to update the `index.json` and `schema.json` files.
```
Please see my comment on `package.json`

This currently doesn't update `index.json`

wolfy1339 replied:
```suggestion
This package updates itself using a daily cronjob running on GitHub Actions. The [`index.json`](index.json) file is generated by [`bin/octokit-webhooks.js`](bin/octokit-webhooks.js). Run `node bin/octokit-webhooks --usage` for instructions. After the update is complete, run `npm run build:schema` to update `schema.json`.
```

wolfy1339 commented on .github/workflows/release.yml:
```diff
@@ -12,6 +12,7 @@ jobs:
       - uses: actions/checkout@v2
       - uses: actions/setup-node@v1
       - run: npm ci
+      - run: npm run build
```
```suggestion
      - run: npm run build:schema
```

wolfy1339 commented on package.json:
```diff
@@ -7,32 +7,36 @@
   },
   "scripts": {
-    "build": "node bin/octokit-webhooks.js check --cached",
+    "build": "npm run build:schema && npm run build:webhooks",
```
```suggestion
    "build": "node bin/octokit-webhooks.js check --cached",
```

G-Rath commented:
@gr2m 

> I would you need de-referencing? Could you use the de-referenced schema as input? That's what we do with the OpenAPI spec -> TypeScript in https://github.com/octokit/openapi-types.ts

"dereference" isn't actually the right word - but in general, the dereferencing being done in the build script is actually bad since it results in duplication and pretty much puts us back where we were in `webhooks.js`.

Currently the single file will give you this massively long union type:

![image](https://user-images.githubusercontent.com/3151613/103103080-195bb500-4684-11eb-8d56-26cdf5438d3a.png)

I don't think this needs to block the PR, as I can make a follow-up with my adjustments that'll correct this and the generate referencing problems.

Effectively what needs to happen is that the `definitions` property for each schema is merged into a single property, and for the `common` schemas to be dereferenced *into definitions*.

We can take advantage of `JSON.stringify`'s `replacer` argument to repoint all `$ref`s point at `common/<interface>.schema.json` with `#/definitions/<interface>`.

Doing this gives us the nice types we want:

![image](https://user-images.githubusercontent.com/3151613/103103917-3e065b80-4689-11eb-9e8f-4b6526ce5e99.png)

gr2m commented:
Maybe I'm missing something, and we can merge the PR and then take it from there

> the dereferencing being done in the build script is actually bad since it results in duplication and pretty much puts us back where we were in `webhooks.js`.

I'd do both, and I wouldn't keep the de-referenced version in the source, only generate it before right before publishing to npm. For generating TypeScript, using the version with references will be much better. The de-referenced version is just a nice to have. But as I said, we can go ahead without. We can add it once there is actually someone who asks for it :)

> here's the patch to `octokit-schema.js` for those wanting to try it out locally

Shall we implement this change in this PR?

G-Rath commented:
Are you maybe thinking of merging? Merging all the schemas into a single `schema.json` definitely has it's usage, but references are a core part of the json schema spec so really there shouldn't be any need for a dereferenced version of the schema.

Overall, dereference is just a fancy way of saying "inlining" i.e

```
{
  "type": "object",
  "definitions": {
    "person": {
      "type": "object",
      "properties": {
        "firstName": { "type": "string" },
        "lastName": { "type": "string" },
        "age": {
          "description": "Age in years",
          "type": "integer",
          "minimum": 0
        }
      },
      "required": ["firstName", "lastName"],
      "additionalProperties": false
    }
  },
  "properties": {
    "people": {
      "type": "array",
      "items": { "$ref": "#/definitions/person" }
    }
  },
  "additionalProperties": false,
  "required": ["people"]
}
```

If you dereference the references in this schema, it'll become:

```
{
  "type": "object",
  "properties": {
    "people": {
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "firstName": { "type": "string" },
          "lastName": { "type": "string" },
          "age": {
            "description": "Age in years",
            "type": "integer",
            "minimum": 0
          }
        },
        "required": ["firstName", "lastName"],
        "additionalProperties": false
      }
    }
  },
  "additionalProperties": false,
  "required": ["people"]
}
```

In this example that's fine because the reference is only used in the one place, but if it's used elsewhere it'll be inlined there too.

Currently the `schema.json` produced is 815 kb - the size of `schema.json` without dereferencing (using `definitions` instead per my patch) is 339 kb (and this'll get even smaller once we move objects to `definitions`, since there's a number of common objects in the events).

While these are under a mb, the difference is still significant - not to mention the fact that dereferencing loses you the ability to make useful interface types.

And then since we've merged everything into a single schema, the references are all completely local so if someone did really want a completely dereferenced version of the schema they could just use `json-schema-ref-parser` like we're doing now :)

-----

> Shall we implement this change in this PR?

I don't mind - I just actually haven't been able to figure out how to push to the branch yet 😅
I think at this point we should just pull the trigger and merge the schemas to make it easier to do & review improvements to them.

wolfy1339 commented:
> I don't mind - I just actually haven't been able to figure out how to push to the branch yet sweat_smile

Here's a little trick to checkout a pull request locally https://gist.github.com/piscisaureus/3342247 though, it's only one-way, meaning you can't push, or it'll push the branch to the repo.

If you use the method I linked,
```
$ git checkout pr/186
<make changes>
$ git commit
$ git push https://github.com/heuels/webhooks pr/186:feature/schema
```

gr2m commented:
okay, let's not worry about the de-referenced schemas for now, it's a problem when it's a problem.

I can do the changes if the method @wolfy1339 shared does not work for you

wolfy1339 commented:
@G-Rath How are you holding up? Did you try to push to the PR branch?

G-Rath commented:
@wolfy1339 sorry no I've not had a chance to get back to this - the office opened back up so been dealing with new year kickoffs and such.

It'll probably be quicker if you or @gr2m push the change, if it's the last thing we want done before merging :)

wolfy1339 commented:
That's no problem. I'll push the patch and we will ship this ASAP :pray:

G-Rath commented:
Awesome - I'll follow it up with a PR for generating the types this weekend 🎉

wolfy1339 commented:
@G-Rath I pushed your patch, I hope I used a commit message that accurately described your patch.

It seems that we are now getting some schema validation errors.

G-Rath commented:
That's because of merging master into the branch, rather than the commit with my changes :)

Looks to be because the recently merged payload examples don't have the `installation` property, which is marked as required by all schemas.

wolfy1339 commented:
I pushed your patch right before I merged in master. 5064c52 is the commit ID

wolfy1339 commented:
@gr2m This is ready to go now

gr2m reviewed (approved):
Great work everyone! This will enable a lot of great work. And I'll let GitHub's Ecosystem API team know, maybe they can help us to further improve the schema and provide more example payloads

gr2m commented:
@G-Rath now that the schema is released, can you update https://github.com/G-Rath/compare-gh-webhook-to-schema-function to use the latest `@octokit/webhooks`? Could you also add some instructions to the README on how to use it? I'd like to share your function with the Ecosystem API team at @github, a bit of documentation on what it is, how it works, and how to set it up / contribute might go a long way to appreciate what you have built :)

G-Rath commented:
@gr2m absolutely! I'll make that my job for the weekend :)

dominguezcelada commented:
I just caught up with this long PR thread. Amazing job @G-Rath and @heuels ❤️ 

This repo as output is simply AWESOME https://github.com/G-Rath/compare-gh-webhook-to-schema-function

I'm willing to hear news from GitHub Team on this :)

gr2m commented:
@G-Rath I'd like to invite you to the @octokit/js-community team, if you like? It will make it easier to contribute to all of JS/TS @octokit repositories. You've been very helpful with webhooks but also the [deno compatibility discussion] and I really appreciate all your help. No pressure of course, and no expectations if you join

G-Rath commented:
@gr2m that would definitely make contributing easier 😅 count me in!

gr2m commented:
I'll take care of it :) It takes a bit longer because @octokit is owned by GitHub, and a special approval is required before adding you to a team. I'll keep you posted :)
